tests/nocover/test_coverage.py

```diff
@@ -0,0 +1,80 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis-python
+#
+# Most of this work is copyright (C) 2013-2017 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import division, print_function, absolute_import
+
+import hypothesis.strategies as st
+from coverage import Coverage
+from hypothesis import given, settings
+from tests.common.utils import all_values
+from hypothesis.database import InMemoryExampleDatabase
+from hypothesis.internal.compat import hrange
+
+
+def test_tracks_and_saves_coverage():
+    db = InMemoryExampleDatabase()
+
+    def do_nothing(): pass
+
+    @settings(database=db)
+    @given(st.integers())
+    def test_branching(i):
+        if i < 0:
+            do_nothing()
+        if i == 0:
+            do_nothing()
+        if i > 0:
+            do_nothing()
+
+    test_branching()
+
+    assert len(all_values(db)) == 3
+
+
+def some_function_to_test(a, b, c):
+    result = 0
+    if a:
+        result += 1
+    if b:
+        result += 1
+    if c:
+        result += 1
+    return result
+
+
+LINE_START = some_function_to_test.__code__.co_firstlineno
+with open(__file__) as i:
```

Comment:
Inspect code objects, then open the current file to get their body as a string?  [Why not!](https://xkcd.com/323/)

(maybe I *should* find this horrifying, but instead I find it hilarious)

Reply:
I actually don't think this is a ballmer peak one. This is more or less the correct thing to do. :-) I mean I *could* hard code these values, but then it wouldn't be robust against editing.

Reply:
Oh, I agree entirely that it's the right thing to do here.

I'm just imagining the reaction of some Java-loving friends of mine... :rofl: